aten/src/ATen/native/TensorAdvancedIndexing.cpp

```diff
@@ -336,6 +336,15 @@ void index_func_meta_impl(
               func, "_(): Number of indices (", numel, ") should be equal to source.size(dim): (",
               source.size(dim), "), for dim: ", dim);
 
+  if (source.dim() != 0) {
+    auto self_sizes = self.sizes().vec();
+    auto source_sizes = source.sizes().vec();
+    self_sizes.erase(self_sizes.begin() + dim);
+    source_sizes.erase(source_sizes.begin() + dim);
+    TORCH_CHECK(!(!self_sizes.empty() && source_sizes.empty()) && at::is_expandable_to(source_sizes, self_sizes),
```

Comment:
I don't think this logic is correct. It looks like it's crafted to make the tested example pass. You are literally asserting that either self is of dimension 1, or sources is of dimension greater than 1 or the sizes of source are NOT broadcastable to the sizes of self...

Reply:
If the error in logic you're identifying is with the check for broadcastability, you shouldn't be applying DeMorgan's law to this portion as it's not in the negated expression. I had to add the negated expression as the error wasn't being thrown for the case self isn't empty and the source is (which I believe the first portion accurately checks).

Reply:
Buf a 0D tensor (which is not empty) is broadcastable to any size, so why would that have to throw an error?

Reply:
(and yes, I misread the parens, sorry for that)

Reply:
> Buf a 0D tensor (which is not empty) is broadcastable to any size, so why would that have to throw an error?

I don't have a great answer, the cop-out is that we require this check to throw the error properly (I understand that this might be reasoning too heavily on abstractions). I think the answer to this has to do with the IntArrayRef representation from implicit construction differing from creating tensors and calling .sizes() on them, as the error was correctly thrown in the latter case. More broadly, perhaps the correct semantics for index_* operations differ slightly from broadcasting? (I'm significantly less sure of this though).

If you have something you'd like me to change the check to I'd be happy to, but I'm not sure what that might be given you'd like me to pass the sizes directly.

Reply:
The semantics of `index_*` operations are the same as that of `x[i] = val`, that is, you should be able to pass a value `val` that is broadcastable to the slice `x[i]`.